File: 3695.maximize-alternating-sum-using-swaps.py
#
# @lc app=leetcode id=3695 lang=python3
#
# [3695] Maximize Alternating Sum Using Swaps
#

# @lc code=start
class Solution:
    def maxAlternatingSum(self, nums: List[int], swaps: List[List[int]]) -> int:
        n = len(nums)
        adj = [[] for _ in range(n)]
        for u, v in swaps:
            adj[u].append(v)
            adj[v].append(u)
        
        visited = [False] * n
        total_sum = 0
        
        for i in range(n):
            if not visited[i]:
                # Start a traversal to find the connected component
                component_indices = []
                stack = [i]
                visited[i] = True
                
                while stack:
                    u = stack.pop()
                    component_indices.append(u)
                    for v in adj[u]:
                        if not visited[v]:
                            visited[v] = True
                            stack.append(v)
                
                # Collect values and count even/odd positions in this component
                values = []
                even_slots = 0
                odd_slots = 0
                
                for idx in component_indices:
                    values.append(nums[idx])
                    if idx % 2 == 0:
                        even_slots += 1
                    else:
                        odd_slots += 1
                
                # Sort values descending
                values.sort(reverse=True)
                
                # Assign largest values to even slots (positive contribution)
                # Assign smallest values to odd slots (negative contribution)
                
                # The first even_slots values are added
                for k in range(even_slots):
                    total_sum += values[k]
                
                # The remaining odd_slots values are subtracted
                for k in range(even_slots, len(values)):
                    total_sum -= values[k]
                    
        return total_sum
# @lc code=end

File: 3695.maximize-alternating-sum-using-swaps.cpp
#
# @lc app=leetcode id=3695 lang=cpp
#
# [3695] Maximize Alternating Sum Using Swaps
#

# @lc code=start
class Solution {
public:
    long long maxAlternatingSum(vector<int>& nums, vector<vector<int>>& swaps) {
        int n = nums.size();
        vector<int> par(n), rnk(n, 0);
        for (int i = 0; i < n; ++i) par[i] = i;
        auto find = [&](auto&& self, int x) -> int {
            return par[x] != x ? par[x] = self(self, par[x]) : x;
        };
        auto union_sets = [&](int a, int b) {
            a = find(find, a);
            b = find(find, b);
            if (a == b) return;
            if (rnk[a] < rnk[b]) swap(a, b);
            par[b] = a;
            if (rnk[a] == rnk[b]) ++rnk[a];
        };
        for (auto& sw : swaps) {
            union_sets(sw[0], sw[1]);
        }
        vector<vector<int>> comp(n);
        for (int i = 0; i < n; ++i) {
            int r = find(find, i);
            comp[r].push_back(i);
        }
        long long ans = 0;
        for (int r = 0; r < n; ++r) {
            if (comp[r].empty()) continue;
            vector<long long> vals;
            int A = 0, B = 0;
            for (int idx : comp[r]) {
                vals.push_back(nums[idx]);
                if ((idx & 1) == 0) ++A;
                else ++B;
            }
            sort(vals.rbegin(), vals.rend());
            int sz = vals.size();
            for (int j = 0; j < A; ++j) ans += vals[j];
            for (int j = A; j < sz; ++j) ans -= vals[j];
        }
        return ans;
    }
};
# @lc code=end